timesource/timesource_test.go

```diff
@@ -175,8 +176,55 @@ func TestNTPTimeSource(t *testing.T) {
 				timeQuery:       tc.query,
 			}
 			assert.WithinDuration(t, time.Now(), source.Now(), clockCompareDelta)
-			source.updateOffset()
+			err := source.updateOffset()
+			if tc.expectError {
+				assert.Equal(t, errUpdateOffset, err)
+			} else {
+				assert.NoError(t, err)
+			}
 			assert.WithinDuration(t, time.Now().Add(tc.expected), source.Now(), clockCompareDelta)
 		})
 	}
 }
+
+func TestRunningPeriodically(t *testing.T) {
+	tc := newTestCases()[0]
+	maxHits := 3
+	minHits := 1
+
+	t.Run(tc.description, func(t *testing.T) {
+		source := &NTPTimeSource{
+			servers:           tc.servers,
+			allowedFailures:   tc.allowedFailures,
+			timeQuery:         tc.query,
+			fastNTPSyncPeriod: time.Duration(maxHits*10) * time.Millisecond,
+			slowNTPSyncPeriod: time.Duration(minHits*10) * time.Millisecond,
+		}
+
+		hits := 0
+		var startingTime time.Time
+		var endingTime time.Time
+
+		err := source.runPeriodically(func() error {
```

Comment:
Consider something like this:
```go
var hits int
var periods []time.Duration

lastCall = time.Now()
// a comment what we simulate here would be nice
err := source.runPeriodically(func() error {
  periods = append(periods, time.Now().Sub(lastCall))
  hits++
  if hits < 3 {
    return errUpdateOffset
  }
  if hits == 6 {
    source.wg.Done()
  }
  return nil
})

source.wg.Wait()
require.NoError(t, err)

require.Len(t, periods, 6)
for _, period := range periods[1:2] {
  // verify it's fast sync
}
for _, period := range periods[2:] {
  // verify it's slow sync
}
```
It was a bit hard for me to follow this logic here. Also, you do not verify the order of periods. It may happen that it switches to the fast sync before the last call and that would not be caught by this test.

Reply:
yeah, looks much clear on your approach. I've replaced existing test, cheers :)